soc/nordic/nrf54h/gpd/gpd.c

```diff
@@ -0,0 +1,229 @@
+	switch (id) {
+	case NRF_GPD_FAST_ACTIVE1:
+		return &fast_active1;
+	case NRF_GPD_SLOW_ACTIVE:
+		return &slow_active;
+	case NRF_GPD_SLOW_MAIN:
+		return &slow_main;
+	default:
+		return NULL;
+	}
+}
+
+static void evt_handler(nrfs_gdpwr_evt_t const *p_evt, void *context)
+{
+	struct gpd_onoff_manager* gpd_mgr = context;
+
+	switch (p_evt->type) {
+	case NRFS_GDPWR_REQ_APPLIED:
+		gpd_mgr->notify(&gpd_mgr->mgr, 0);
+		break;
+	default:
+		gpd_mgr->notify(&gpd_mgr->mgr, -EIO);
+		break;
+	}
+}
+
+static void start(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	nrfs_err_t err;
+	struct gpd_onoff_manager* gpd_mgr =
+		CONTAINER_OF(mgr, struct gpd_onoff_manager, mgr);
+
+	gpd_mgr->notify = notify;
+
+	err = nrfs_gdpwr_power_request(gpd_mgr->id, GDPWR_POWER_REQUEST_SET, gpd_mgr);
+	if (err != NRFS_SUCCESS) {
+		notify(mgr, -EIO);
+	}
+}
+
+static void stop(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	nrfs_err_t err;
+	struct gpd_onoff_manager* gpd_mgr =
+		CONTAINER_OF(mgr, struct gpd_onoff_manager, mgr);
+
+	gpd_mgr->notify = notify;
+
+	err = nrfs_gdpwr_power_request(gpd_mgr->id, GDPWR_POWER_REQUEST_CLEAR, gpd_mgr);
+	if (err != NRFS_SUCCESS) {
+		notify(mgr, -EIO);
+	}
+}
+
+int nrf_gpd_request(uint8_t id)
+{
+	if (atomic_test_bit(&gpd_service_status, GPD_SERVICE_READY)) {
+		int ret;
+		struct onoff_client client;
+		struct gpd_onoff_manager* gpd_mgr;
+
+		gpd_mgr = get_mgr(id);
+		if (gpd_mgr == NULL) {
+			return -EINVAL;
+		}
+
+		sys_notify_init_spinwait(&client.notify);
+
+		onoff_request(&gpd_mgr->mgr, &client);
+
+		while (sys_notify_fetch_result(&client.notify, &ret) == -EAGAIN) {}
+
+		return ret;
+	} else {
+		atomic_t *gpd_cnt;
+
+		gpd_cnt = get_cnt(id);
+		if (gpd_cnt == NULL) {
+			return -EINVAL;
+		}
+
+		(void)atomic_inc(gpd_cnt);
+
+		return 0;
+	}
+}
+
+int nrf_gpd_release(uint8_t id)
+{
+	if (atomic_test_bit(&gpd_service_status, GPD_SERVICE_READY)) {
+		struct gpd_onoff_manager* gpd_mgr;
+
+		gpd_mgr = get_mgr(id);
+		if (gpd_mgr == NULL) {
+			return -EINVAL;
+		}
+
+		return onoff_release(&gpd_mgr->mgr);
+	} else {
+		atomic_t *gpd_cnt;
+
+		gpd_cnt = get_cnt(id);
+		if (gpd_cnt == NULL) {
+			return -EINVAL;
+		}
+
+		(void)atomic_dec(gpd_cnt);
+
+		return 0;
+	}
+}
+
+static int nrf_gpd_init(void)
+{
+	nrfs_err_t err;
+	int ret;
+
+	ret = nrfs_backend_wait_for_connection(K_NO_WAIT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	err = nrfs_gdpwr_init(evt_handler);
+	if (err != NRFS_SUCCESS) {
+		return -EIO;
+	}
+
+	ret = onoff_manager_init(&fast_active1.mgr, &transitions);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = onoff_manager_init(&slow_active.mgr, &transitions);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = onoff_manager_init(&slow_main.mgr, &transitions);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* flag GPD service as ready */
+	atomic_set_bit(&gpd_service_status, GPD_SERVICE_READY);
+
+	/* release any domain that was not left active during the boot process */
```

Comment:
The double reference counting between "before GPD_SERVICE_READY and after GPD_SERVICE_READY" isn't correct, you'd need to inject the different reference counters such as `fast_active1_cnt` into the respective onoff manager by recalling `nrf_gpd_request()` multiple times to make sure all the positive requests are still there, otherwise when something turns off again later that was turned on but not tracked it might lead to an invalid off transition.

Or drop the double reference counting and just rely on onoff manager. In the start/stop transitions you can check if the nrfs backend is available. That means you'd need two init stages for the gpd, one early to initialize the onoff managers and one late to send the state after nrfs is connected and send any state that wasn't transmitted yet